cmake/CPM.cmake

```diff
@@ -464,6 +464,80 @@ function(cpm_check_git_working_dir_is_clean repoPath gitTag isClean)
 
 endfunction()
 
+# method to add PATCH_COMMAND to CPM_ARGS_UNPARSED_ARGUMENTS for PATCHES files.
+function(cpm_add_patches)
+  # Return if no patch files are supplied.
+  if(NOT ARGN)
+    return()
+  endif()
+
+  # Provide a small warning.
+  if(CMAKE_VERSION VERSION_LESS "3.20")
+    message(
+      AUTHOR_WARNING
+        "Versions of CMake less than 3.20 may need to supply patch files as absolute paths."
+    )
+  endif()
+
+  # Find the patch program.
+  find_program(PATCH_EXECUTABLE patch)
+  if(WIN32 AND NOT PATCH_EXECUTABLE)
+    # The Windows git executable is distributed with patch.exe. Find the path to the executable, if
+    # it exists, then search `../../usr/bin` for patch.exe.
+    find_package(Git QUIET)
+    if(GIT_EXECUTABLE)
+      if(CMAKE_VERSION VERSION_GREATER_EQUAL "3.20")
+        cmake_path(GET GIT_EXECUTABLE PARENT_PATH extra_search_path)
+        cmake_path(GET extra_search_path PARENT_PATH extra_search_path)
+      else()
+        get_filename_component(extra_search_path ${GIT_EXECUTABLE} DIRECTORY)
+        get_filename_component(extra_search_path ${extra_search_path} DIRECTORY)
+        get_filename_component(extra_search_path ${extra_search_path} DIRECTORY)
+      endif()
+      find_program(PATCH_EXECUTABLE patch HINTS "${extra_search_path}/usr/bin")
+    endif()
+  endif()
+  if(NOT PATCH_EXECUTABLE)
+    message(FATAL_ERROR "Couldn't find `patch` executable to use with PATCHES keyword.")
+  endif()
+
+  # Create a temporary
+  set(temp_list ${CPM_ARGS_UNPARSED_ARGUMENTS})
+
+  # Ensure each file exists (or error out) and add it to the list.
+  set(first_item True)
+  foreach(PATCH_FILE ${ARGN})
+    if(NOT EXISTS "${PATCH_FILE}")
+      if(NOT EXISTS "${CMAKE_CURRENT_LIST_DIR}/${PATCH_FILE}")
+        message(FATAL_ERROR "Couldn't find patch file: '${PATCH_FILE}'")
+      endif()
+      set(PATCH_FILE "${CMAKE_CURRENT_LIST_DIR}/${PATCH_FILE}")
+    endif()
+    # Convert to absolute path for CMake 3.20 (available since March 3rd, 2021) and later, this may
+    # cause users with older versions of CMake to strictly call out an absolute path.
+    if(CMAKE_VERSION VERSION_GREATER_EQUAL "3.20")
+      cmake_path(ABSOLUTE_PATH PATCH_FILE)
+    endif()
+    # The first patch entry must be preceded by "PATCH_COMMAND" while the following items are
+    # preceded by "&&".
+    if(first_item)
+      set(first_item False)
+      list(APPEND temp_list "PATCH_COMMAND")
+    else()
+      list(APPEND temp_list "&&")
+    endif()
+    # Add the patch command to the list
+    list(APPEND temp_list "${PATCH_EXECUTABLE}" "-p1" "<" "${PATCH_FILE}")
+  endforeach()
+
+  # Move temp out into parent scope.
+  set(CPM_ARGS_UNPARSED_ARGUMENTS
+      ${temp_list}
+      PARENT_SCOPE
+  )
```

Comment:
I'm not sure I understand the suggestion? Can you point me to an example?

In this case, we put `cpm_add_patches` into a function to keep the code more readable. I suppose we could send `CPM_ARGS_UNPARSED_ARGUMENTS` in as an argument, but I feel like the documentation would always be `VARIABLE_TO_UPDATE _must_ _be_ CPM_ARGS_UNPARSED_ARGUMENTS`. This could be more confusing overall, I think.

Reply:
@TheLartians 523c14a1c8b2f9625c0d5c55d2d3deb3c529da74 addresses the other comments, but I'm not sure what to do here... Suggestions? Thanks!

Reply:
I see your point that the function is basically just an outsourced section of a larger function, but in general I like to have clearly defined input and outputs in functions - in CMake's case using arguments so the caller can decide what goes in and comes out. In my experience ist helps a lot with readability. In any way it's just a style preference and not a blocker for this PR.